### PR TITLE
feat: Normalised MPI and BHR + made EIP-5164 compliant

### PR DESCRIPTION
**Pull request by allemanfredi**
As described in the title, this pr proposes some changes in order for Hashi to be compliant with [EIP-5164](https://eips.ethereum.org/EIPS/eip-5164).

## TODO
- [x] Normalized all `messageIds` types to `bytes32`
- [x] Changed the function interfaces within `Yaho` according to the EIP.
- [x] Changed how `messageId` and `messageHash` are calculated.

```js
salt = isHeaderReporter ? keccak256(blockNumber, MESSAGE_BHR) : keccak256(blockhash(block.number), gasleft(), MESSAGE_MPI)
messageId = keccak256(fromChainId, yaho, salt)
messageHash = keccak256(message, messageId, yahoAddress)
```

- [x] Changed how `Yaru` invokes the target contract. In order for it to be called, it must implement the following function:
```sol
function onMessage(bytes calldata data, bytes32 messageId, uint256 fromChainId, address from) external returns (bytes memory);
```
- [x] Added `toChainIds` to `IMessageRelay.relayMessages`.
- [x] Remove ambiguity between `messageIds` calculation within `MessageRelay` and `HeaderReporter` contracts.
- [x] Added `HeaderVault`.
- [x] Deleted `NativeTokenBridge`.
- [ ] Added commitment to used adapters within the messageHash in order to don't execute a message with a not selected adapter

### CODE REVIEW

**Comment by allemanfredi** [comment]
We cannot merge the pr because at the moment, since message relayers does not allow to specify the destination chain id. For this reason it would be possible to execute a message where `message.toChainId != block.chainid`

**Comment by allemanfredi** [comment]
> 03942a7

Solved in [03942a7](https://github.com/gnosis/hashi/pull/28/commits/03942a7e6f3ca97329ec70f84db022bee333e468)!

**Comment by allemanfredi** [comment]
> We cannot merge the pr because at the moment, since message relayers does not allow to specify the destination chain id. For this reason it would be possible to execute a message where `message.toChainId != block.chainid`

Solved in [03942a7](https://github.com/gnosis/hashi/pull/28/commits/03942a7e6f3ca97329ec70f84db022bee333e468)!

File: packages/evm/contracts/Yaho.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/Yaho.sol:36]: Why overload `dispatchMessages()`?
It seems the version below, where `data` is a array of bytes, would almost always be preferable.

File: packages/evm/contracts/Yaho.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/Yaho.sol:22]: > putting their **??** into storage

Looks like an old typo. Should say *"putting their hashes into storage"*.

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/Yaru.sol:4]: `Yaru` was not intended to be `Ownable`.

Similar to `Hashi`, I'd prefer that an ownable version of this be a contract that either inherits or reads from this contract.

Two possible solutions that don't require the contract to be ownable:

1. Have a separate `Yaru` deployment for each network.
2. Assume `Yaho` is always deployed to the same address on all networks and leave it to the adapters to specify the network ID that the message came from.

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/Yaru.sol:5]: I generally prefer not to use ReentrancyGuard, and instead to write code to specifically thwart reentrancy for the given scenario.

Do you see any issues with the previous code to prevent reentrancy?

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/Yaru.sol:14]: `Yaru` should be use-case agnostic, so it should refer to "hashes" rather than "headers".

I'd prefer that there not be something like `HeaderVault` enshrined at the `Yaru` layer. It should be a layer built on top, as per the above comment about `Ownable`.

It seems to me that this functionality belongs in a contract in the `ownable` folder.

**Review by auryn-macmillan — CHANGES_REQUESTED** [review]

File: packages/evm/contracts/test/NaiveTokenBridge.sol
Comment (auryn-macmillan) [inline packages/evm/contracts/test/NaiveTokenBridge.sol:1]: Why delete this?

File: packages/evm/contracts/Yaru.sol
Comment (bertani) [reply packages/evm/contracts/Yaru.sol:4]: I agree with you in principle @auryn-macmillan, however note that once initialized it cannot get changed (because of the initializeForChainId method).
Given that, option 1. would imply the same trust model (you verify before deployment what the address is, and that it is initialized correctly) while duplicating code and overcomplicating the development process for Hashi users, while 2. doesn't seem to be something guaranteed to work on all networks (i.e. zksync?).

I still feel like this approach is a better balance. Ownable is just used to avoid anyone to mess up with yet uninitialized network references (as new networks could be added later), but doesn't really require trust as long as you verify before deploying your code that the network you want to use has indeed been added to Hashi and safely initiatlized. Once set, it cannot be changed by anyone, similarly to any initializable code.

File: packages/evm/contracts/Yaho.sol
Comment (allemanfredi) [reply packages/evm/contracts/Yaho.sol:36]: because if you dispatch the same data on multiple chains it is as if you generate multiple messages

File: packages/evm/contracts/Yaru.sol
Comment (allemanfredi) [reply packages/evm/contracts/Yaru.sol:5]: No but since we removed `sender`, i decided to use the oz one!

File: packages/evm/contracts/test/NaiveTokenBridge.sol
Comment (allemanfredi) [reply packages/evm/contracts/test/NaiveTokenBridge.sol:1]: error sorry!

File: packages/evm/contracts/Yaho.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaho.sol:22]: Worth noting that this typo exists on all of the derivatives of this function.

File: packages/evm/contracts/Yaho.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaho.sol:36]: Right, I get that it goes out as multiple messages.

I'm mostly wondering what the use-case here is and whether it's common enough to justify a separate entrypoint? In other words, how often would one send the exact same data to many different destinations?

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:4]: Right, but using `Ownable`, however small, now enshrines some entity as THE thing that can make canonical changes to the base layer, which is explicitly counter to the original goal of hashi.

Another way to solve for this might be to offload this to the `IOracleAdapter`. Essentially, each oracle adapter would define what it views as the cannonical `Yaho` deployment for any given network.

File: packages/evm/contracts/Yaho.sol
Comment (allemanfredi) [reply packages/evm/contracts/Yaho.sol:36]: For example if you wanted to propagate a block from a source chain to many destionation chains since now blocks propagation passes through `Yaho` in order to correctly calculate the messageId based on the EIP-5164 standard

File: packages/evm/contracts/Yaru.sol
Comment (allemanfredi) [reply packages/evm/contracts/Yaru.sol:14]: Let me explain. 

With the introduction of EIP-5164, the old messageId calculation has become obsolete. This is because using block numbers as if they were message IDs is not compliant, and there is also the possibility of having messages with the same message ID. For this reason, we have decided to always use `Yaho` for sending messages, whether it's for block header reporting or normal message passing.

Regarding the reason why we decided to add a link to `HeaderVault` within `Yaru`, it is because we need a way to natively distinguish it from normal contracts. In case we wanted to remove `HeaderVault` we could just revert if `message.from = address(0) && message.to = address(0)` (blocks reporting).

This approach greatly simplifies the flow. We will essentially have a unique process for sending both blocks and messages. External oracles will only need to write the `MessageRelay` and Adapter contracts, eliminating the need for a `HeaderReporter`. Moreover, with this approach, it is always possible to fetch a block hash from `Hashi` given the block number, ensuring compatibility with the standard. Ex:

```
function getBlockHeadersFromOracles(
        IOracleAdapter[] calldata oracles,
        uint256 fromChainId,
        uint256 blockNumber
    ) external view returns (bytes32[] memory) {
        return
            IHashi(hashi).getHashesFromOracles(
                oracles,
                fromChainId,
                calculateMessageId(
                    fromChainId,
                    block.chainid,
                    IYaru(yaru).yahos(fromChainId),
                    address(0),
                    address(0),
                    keccak256(abi.encode(blockNumber)),
                    MESSAGE_BHR
                )
            );
    }
```

File: packages/evm/contracts/Yaho.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaho.sol:36]: Ah yeah, that's fair enough justification for a distinct entrypoint.

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:5]: Right, but what did you gain with this change?

You essentially swapped this:
```solidity
    address public sender;
```
for this:
```solidity
    uint256 private constant _NOT_ENTERED = 1;
    uint256 private constant _ENTERED = 2;

    uint256 private _status;
```

If there were other functions marked `nonReentrant`, then the change might be more justified. But I don't see any good reason for it as is, unless there is some vulnerability in the previous code.

File: packages/evm/contracts/Yaru.sol
Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: Wait... why relay blocks through `Yaru`? That seems needlessly complicated?
The whole point of Hashi is to be modular. `Yaru` can be EIP-5164 compliant, and be used for passing messages (which may contain block headers in some cases), but there is no reason that all routes should be forced to use `Yaru`.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: > there is also the possibility of having messages with the same message ID

This is not an issue, since it's up to the respective oracle adapters to determine what IDs mean.

Comment (allemanfredi) [reply packages/evm/contracts/Yaru.sol:14]: it's not really like that! let's say that the blocks might not even pass through `Yaru` since you would still have the hash saved in each adapter. The idea of ​​the `HeaderVault` was to natively manage the execution of a message containing a block header. A possible solution is to reverse the execution of those messages.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: > A possible solution is to reverse the execution of those messages.

Not sure what you mean here. Could you elaborate some?

> The idea of ​​the HeaderVault was to natively manage the execution of a message containing a block header.

This is exactly my point though. I don't think this should be native to `Yaru`, rather it should be a separate contract that `Yaru` passes a message to.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: Hashi is constructed as a collection of distinct and useful "stages". This is a really valuable feature from the point of view of composability. Wherever possible, we should resist the urge to bundle things together / natively support specific use-cases and the lower/more generalized stages.

Comment (allemanfredi) [reply packages/evm/contracts/Yaru.sol:14]: Since now block header messages and normal ones have the same format (same `Message` struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages. 

I understand your point of building `HeaderVault` on top of `Yaru` but in that case, however, the propagation of the blocks would no longer be native because it could be done by everyone.


> > A possible solution is to reverse the execution of those messages.
> 
> Not sure what you mean here. Could you elaborate some?
> 
> > The idea of ​​the HeaderVault was to natively manage the execution of a message containing a block header.
> 
> This is exactly my point though. I don't think this should be native to `Yaru`, rather it should be a separate contract that `Yaru` passes a message to.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: > the propagation of the blocks would no longer be native because it could be done by everyone.

I don't see why this is the case. The HeaderVault would simply validate the sending address.

> Since now block header messages and normal ones have the same format (same Message struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages.

If `HeaderVault` is a separate contract, then the "execution" of a message is simply the call to `HeaderVault`. In other words, there is no need for any special handling in this case.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: This is functionally similar to the existing `HeaderStorage` contract.

Comment (allemanfredi) [reply packages/evm/contracts/Yaru.sol:14]: > > the propagation of the blocks would no longer be native because it could be done by everyone.
> 
> I don't see why this is the case. The HeaderVault would simply validate the sending address.
> 
> > Since now block header messages and normal ones have the same format (same Message struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages.
> 
> If `HeaderVault` is a separate contract, then the "execution" of a message is simply the call to `HeaderVault`. In other words, there is no need for any special handling in this case.

Correct! But then in order to access the block header hash within `Hashi` you would need also the `message.from` and the `message.to`, which is not intuitive and as far as i know fetching a block hash using the block number was a requirement.

Comment (auryn-macmillan) [reply packages/evm/contracts/Yaru.sol:14]: This is the job of the adapters though.
Like if my adapter is using Yaru, then it's my adapter's job to know who the sender should be.

See the AMB adapter as a reference for this pattern.

**Comment by allemanfredi** [comment]
Closed this as it is stale

**Pull request closed without merging**
